UnitySDK/Assets/ML-Agents/Scripts/InferenceBrain/TensorGenerator.cs

```diff
@@ -0,0 +1,240 @@
+            tensor.Shape[0] = batchSize;
+            var memorySize = tensor.Shape[1];
+            tensor.Data = new float[batchSize, memorySize];
+            var agentIndex = 0;
+            foreach (var agent in agentInfo.Keys)
+            {
+                var memory = agentInfo[agent].memories;
+                for (var j = 0; j < memorySize; j++)
+                {
+                    if (memory == null)
+                    {
+                        break;
+                    }
+
+                    if (j >= memory.Count)
+                    {
+                        break;
+                    }
+                    tensor.Data.SetValue(memory[j], new int[2] {agentIndex, j});
+                }
+                agentIndex++;
+            }
+        }
+    }
+
+    /// <summary>
+    /// Generates the Tensor corresponding to the Previous Action input : Will be a two
+    /// dimensional integer array of dimension [batchSize x actionSize].
+    /// It will use the previous action data contained in the agentInfo to fill the data
+    /// of the tensor.
+    /// </summary>
+    public class PreviousActionInputGenerator : TensorGenerator
+    {
+        public void Execute(Tensor tensor, int batchSize, Dictionary<Agent, AgentInfo> agentInfo)
+        {
+            if (tensor.ValueType != Tensor.TensorType.Integer)
+            {
+                throw new NotImplementedException(
+                    "Previous Action Inputs are only valid for discrete control");
+            }
+
+            tensor.Shape[0] = batchSize;
+            var actionSize = tensor.Shape[1];
+            tensor.Data = new int[batchSize, actionSize];
+            var agentIndex = 0;
+            foreach (var agent in agentInfo.Keys)
+            {
+                var pastAction = agentInfo[agent].storedVectorActions;
+                for (var j = 0; j < actionSize; j++)
+                {
+                    tensor.Data.SetValue((int) pastAction[j], new int[2] {agentIndex, j});
+                }
+
+                agentIndex++;
+            }
+        }
+    }
+
+    /// <summary>
+    /// Generates the Tensor corresponding to the Action Mask input : Will be a two
+    /// dimensional float array of dimension [batchSize x numActionLogits].
+    /// It will use the Action Mask data contained in the agentInfo to fill the data
+    /// of the tensor.
+    /// </summary>
+    public class ActionMaskInputGenerator : TensorGenerator
+    {
+        public void Execute(Tensor tensor, int batchSize, Dictionary<Agent, AgentInfo> agentInfo)
+        {
+            tensor.Shape[0] = batchSize;
+            var maskSize = tensor.Shape[1];
+            tensor.Data = new float[batchSize, maskSize];
+            var agentIndex = 0;
+            foreach (var agent in agentInfo.Keys)
+            {
+                var maskList = agentInfo[agent].actionMasks;
+                for (var j = 0; j < maskSize; j++)
+                {
+                    if (maskList != null)
```

Comment:
couldn't we ignore the if-else and replace with:
`tensor.Data.SetValue((maskList != null && maskList[j]) ? 0.0f : 1.0f, new int[2] {agentIndex, j});` or better, move `(maskList != null && maskList[j]) ? 0.0f : 1.0f` to a temp variable